third_party/move/move-compiler-v2/src/env_pipeline/unused_params_checker.rs

```diff
@@ -55,12 +55,16 @@ fn used_type_parameters_in_fields(struct_env: &StructEnv) -> BTreeSet<u16> {
 fn used_type_parameters_in_ty(ty: &Type) -> BTreeSet<u16> {
     match ty {
         Type::Primitive(_) => BTreeSet::new(),
-        Type::Struct(_, _, tys) => tys.iter().flat_map(used_type_parameters_in_ty).collect(),
+        Type::Tuple(tys) | Type::Struct(_, _, tys) => {
+            tys.iter().flat_map(used_type_parameters_in_ty).collect()
+        },
         Type::TypeParameter(i) => BTreeSet::from([*i]),
         Type::Vector(ty) => used_type_parameters_in_ty(ty),
+        Type::Fun(t1, t2) => [t1, t2]
```

Comment:
This may not be the right behavior:

Here, we detect unused (non-phantom) type params and suggest that they be made phantom. But the ability requirement enforcement may be different for struct `S<T>` that has a field with type `T`, vs. a struct `S<T>` that has a field with type `|T|` (i.e., a function that consumes an `T` and returns nothing). That is, we could still make `T` a phantom type, because it is not "directly" used in making up the type's representation.

Reply:
edit: after thinking about this again, I realized this should be okay, because we are complaining about unused type params here and not identifying ability requirements, which I was conflating in the write up above.